### DISCUSSION TITLE
Are Fully featured web applications on the horizon for NiceGUI?

### DISCUSSION THREAD

**Discussion by metaperl** [zauberzeug/nicegui#135] [Q&A] [answered]
Hello, I've written [a review of NiceGUI](https://may69.com/purepython/#NiceGUI) and I may have the wrong idea in some places. Regarding authentication and authorization, I have written the following:

> cannot produce fully featured web applications. It has all the other elements of a Class A solution. There is [a toy example of authentication](https://github.com/zauberzeug/nicegui/blob/main/examples/authentication/main.py) but it is not ready to use: 
1. they did not show how to use authorization groups to permit/deny access to web application resources. 
2.  their code does not gracefully use various storage options.
3. They do not show examples of trying to access a URL and being redirected to the login page because of lack of authorization
4. They did not show how to store authentication credentials to allow direct access to a resource previously authorized.

**Comment by me21** [comment]
Yes, NiceGUI currently does not touch these features - it's a GUI library (narrower scope), not a web framework. Hopefully one day someone will write some plugins (nicegui-auth, anyone?).
That said, I'm a user of this library, not its developer. Maybe they have something in their roadmaps.

**Comment by falkoschindler** [comment]
@me21 is right: NiceGUI is primarily a library for building GUIs in Python. By rendering to the browser we not only get network access for free, but can use mighty web frameworks like Vue and Quasar as well. The result is a library that almost feels like a "fully featured" web application. And maybe - depending on the definition - it already is. The underlying Uvicorn and Starlette already give plenty possibilities.

But we have indeed plans for moving further towards web applications. We're currently experimenting with a core implementation without JustPy to remove one major indirection and make us more independent in our choice of central building blocks. E.g. we're thinking about using a FastAPI server instead of plain Starlette. Besides that we will automatically upgrade from Vue 2 to Vue 3 and Quasar 1 to Quasar 2, making NiceGUI more future-proof.

**Reply by metaperl** [reply]
> we're thinking about using a FastAPI server instead of plain Starlette.

Have you considered [Starlite](https://github.com/starlite-api/starlite)? They had some very [real reasons](https://starlite-api.github.io/starlite/#about-the-starlite-project) for not using FastAPI or Starlette.

**Reply by me21** [reply]
As they say themselves, Starlite is built on top of Starlette.

**Reply by rodja** [reply]
@metaperl Thanks for pointing out Starlite as an alternative. We where not aware of its existence. After some considerations we very possibly stick with FastAPI as a new basis for a few reasons:

- it's much more well-known (right now)
- the framework is less opinionated than Starlite which makes it easier for us to implement our ideas we have with NiceGUI
- we have gained a lot of experience with FastAPI over the last years and are very happy with it's reliability

This may change in the future. We are not bound to the underlying framework forever. It's crucial to revisit alternatives regularly and identify the best solutions for our requirements.

**Comment by rodja** [comment]
Right now NiceGUI is limited by some design choices in JustPy. For example the "catch all" route stops us from applying the standard Starlette middleware for authentication. As you mentioned in your very good review this is currently an obstacle to become a full featured web framework. But stay tuned. The future is bright. I'm already grinning whenever I think about the stuff we are currently cooking.

**Comment by rodja** [comment, marked as answer]
@metaperl we just released NiceGUI 1.0: https://github.com/zauberzeug/nicegui/releases/tag/v1.0.0  🎉

Now NiceGUI is build on top of FastAPI which allows you to write fully featured web applications. Please have a look and tell us what you think.

**Reply by metaperl** [reply]
yes, I've moved NiceGUI to Class A - https://may69.com/purepython/#Class_A_Industrial_Strength_Rock-solid_Products

**Reply by rodja** [reply]
Thanks @metaperl. Maybe you can also revisit your section about "name" Attributes? See https://www.reddit.com/r/Python/comments/10d6ugv/comment/j4l3sdx/?utm_source=reddit&utm_medium=web2x&context=3
